src/main/java/uk/gov/companieshouse/overseasentitiesapi/utils/PublicPrivateDataCombiner.java

```diff
@@ -3,258 +3,280 @@
+
+    private void createOELog(StringBuilder result) {
+        result.append("\"Overseas Entity Data\": {");
+        if (this.combinedOEs != null && this.combinedOEs.getLeft() != null && this.combinedOEs.getRight() != null) {
+            result.append("\"Public Entity Number\": \"" + this.combinedOEs.getLeft().getCompanyNumber() + "\", ");
+            result.append("\"Private Email\": \"" + this.combinedOEs.getRight().getEmail() + "\"");
+        }
+        result.append("}");
     }
-    result.append("}");
-  }
 
+    private void putPrivateBoData(PrivateBoDataListApi privateBOs) throws ServiceException {
+        for (PrivateBoDataApi privateBO : privateBOs.getBoPrivateData()) {
+            String hashedId = retrieveHashedId(privateBO.getPscId());
 
-  public String logCollatedData() {
-    var result = new StringBuilder("{");
+            var pairFromMap = combinedBOs.get(hashedId);
+            Pair<PscApi, PrivateBoDataApi> pairToPutInMap;
 
-    createMoLog(result);
-    createBoLog(result);
-    createOELog(result);
+            if (pairFromMap != null) {
+                pairToPutInMap = Pair.of(pairFromMap.getLeft(), privateBO);
+            } else {
+                pairToPutInMap = Pair.of(null, privateBO);
+            }
+            combinedBOs.put(hashedId, pairToPutInMap);
+        }
+    }
 
-    result.append("}");
+    private void putPublicBoData(PscsApi publicPSCs) {
+        for (PscApi publicPSC : publicPSCs.getItems()) {
+            String[] linkAsArray = publicPSC
+                    .getLinks()
+                    .getSelf()
+                    .split("/");
+            String hashedId = linkAsArray[linkAsArray.length - 1];
+
+            var pairFromMap = combinedBOs.get(hashedId);
+            Pair<PscApi, PrivateBoDataApi> pairToGoInMap;
+
+            if (pairFromMap != null) {
+                pairToGoInMap = Pair.of(publicPSC, pairFromMap.getRight());
+            } else {
+                pairToGoInMap = Pair.of(publicPSC, null);
+            }
+            combinedBOs.put(hashedId, pairToGoInMap);
+        }
+    }
 
-    return result.toString();
-  }
+    private void putPrivateMoData(ManagingOfficerListDataApi privateMOs) throws ServiceException {
+        for (ManagingOfficerDataApi privateMO : privateMOs.getManagingOfficerData()) {
+            String hashedId = retrieveHashedId(privateMO.getManagingOfficerId());
+
+            var pairFromMap = combinedMOs.get(hashedId);
+            Pair<CompanyOfficerApi, ManagingOfficerDataApi> pairToPutInMap;
+
+            if (pairFromMap != null) {
+                pairToPutInMap = Pair.of(pairFromMap.getLeft(), privateMO);
+            } else {
+                pairToPutInMap = Pair.of(null, privateMO);
+            }
+
+            combinedMOs.put(hashedId, pairToPutInMap);
+        }
+    }
+
+    private void putPublicMoData(OfficersApi publicOfficers) {
+
+        for (CompanyOfficerApi publicOfficer : publicOfficers.getItems()) {
```

Comment:
Can items be null? The exception was on this line but I'm not sure if it was due to `publicOfficers` being null or `getItems` returning null

Reply:
Looking at PublicDataRetrievalService.java, line 86, if the officer data cannot be found for a company number we return an `officersApi` object but only set the `totalResults` so the `items` data may be null.